File: scripts/storyline.py
# scripts/storyline.py
# Простая сюжетная линия: уровни, диалоги, цели, события

class StoryLevel:
    def __init__(self, level_id, title, dialog, objective, reward=None):
        self.level_id = level_id
        self.title = title
        self.dialog = dialog  # текст, который показывается при входе
        self.objective = objective  # что нужно сделать
        self.reward = reward or []  # список наград (имена перков/бонусов/скинов)
        self.completed = False

    def check_complete(self, player, game_state):
        """Проверяет, выполнена ли цель уровня (переопределяй в подклассах)"""
        return False

    def on_complete(self, player):
        """Что происходит при завершении уровня"""
        self.completed = True
        print(f"Уровень {self.title} завершён!")
        # Здесь можно добавить перки, бонусы, скины и т.д.
        for rew in self.reward:
            print(f"Получена награда: {rew}")
            # Пример: if rew == "double_jump": player.double_jump = True


# Список уровней сюжета (можно расширять бесконечно)
STORY_LEVELS = [
    StoryLevel(
        level_id=1,
        title="Пробуждение",
        dialog="Ты просыпаешься в тёмной комнате. Где-то вдалеке слышен голос: 'Найди ключ к двери...'",
        objective="Найди и подбери ключ (нажми E рядом с ним)",
        reward=["Быстрее ноги", "Скоростной режим"]
    ),
    
    StoryLevel(
        level_id=2,
        title="Первый враг",
        dialog="Перед тобой появляется странное существо. Оно агрессивно... Придётся сражаться.",
        objective="Победить первого врага (пока просто коснись его)",
        reward=["Железная кожа"]
    ),
    
    StoryLevel(
        level_id=3,
        title="Тайный проход",
        dialog="Дверь открылась, но за ней — лабиринт. Собери 3 кристалла, чтобы открыть портал.",
        objective="Собрать 3 кристалла",
        reward=["Двойной прыжок", "Мега-прыжок"]
    ),
    
    StoryLevel(
        level_id=4,
        title="Босс тени",
        dialog="Ты дошёл до конца... Но тень ждёт тебя. Это финал этой главы.",
        objective="Победить босса (пока просто продержись 30 секунд)",
        reward=["Невидимость на 5 сек"]
    )
]


class StoryManager:
    def __init__(self):
        self.current_level_id = 1
        self.levels = {lvl.level_id: lvl for lvl in STORY_LEVELS}
        self.current_dialog_shown = False

    def get_current_level(self):
        return self.levels.get(self.current_level_id)

    def advance_level(self, player):
        if self.current_level_id < len(STORY_LEVELS):
            self.current_level_id += 1
            self.current_dialog_shown = False
            print(f"Переход на уровень {self.current_level_id}")
            return True
        print("Сюжет завершён!")
        return False

    def show_dialog(self, screen, font):
        level = self.get_current_level()
        if level and not self.current_dialog_shown:
            # Рисуем диалог внизу экрана
            dialog_surf = font.render(level.dialog, True, (255, 220, 100))
            title_surf = font.render(level.title, True, (200, 100, 255))
            screen.blit(title_surf, (20, HEIGHT - 120))
            screen.blit(dialog_surf, (20, HEIGHT - 80))
            # Чтобы не спамило каждый кадр — показываем один раз
            self.current_dialog_shown = True

    def check_objective(self, player, game_state):
        level = self.get_current_level()
        if level and level.check_complete(player, game_state):
            level.on_complete(player)
            self.advance_level(player)


# Глобальный менеджер сюжета (создавай один экземпляр в main.py)
story_manager = StoryManager()

File: js/storyline.js
// js/storyline.js
// Сюжетная линия — 50+ уровней, диалоги, цели, награды, ветвления

const STORY_LEVELS = [
  { id: 1, title: "Пробуждение", dialog: "Ты открываешь глаза в древнем храме. Голос шепчет: 'Собери осколки света, чтобы открыть путь...'", objective: "Собери 5 монет", reward: ["Speed Demon"], branch: null },
  { id: 2, title: "Первые тени", dialog: "Тени оживают. Они хотят твою душу. Уничтожь их!", objective: "Убей 3 врага", reward: ["Invincibility Burst"], branch: null },
  { id: 3, title: "Лабиринт забвения", dialog: "Стены шепчут ложь. Найди истинный путь.", objective: "Дойди до конца лабиринта", reward: ["Double Jump", "Magnet Coins"], branch: null },
  { id: 4, title: "Огненный страж", dialog: "Огонь испытывает тебя. Выдержишь ли?", objective: "Переживи 20 сек в огне", reward: ["Fire Aura"], branch: null },
  { id: 5, title: "Ледяная бездна", dialog: "Холод замораживает кровь. Разбей лед.", objective: "Разбей 10 ледяных блоков", reward: ["Ice Shield"], branch: null },
  { id: 6, title: "Теневое предательство", dialog: "Твой союзник — тень. Он предаёт.", objective: "Победить предателя", reward: ["Shadow Teleport"], branch: null },
  { id: 7, title: "Портал хаоса", dialog: "Порталы ведут в неизвестность. Выбери правильно.", objective: "Найди правильный портал", reward: ["Portal Jump"], branch: "good" },
  { id: 8, title: "Добрый путь", dialog: "Ты выбрал свет. Теперь мир благодарен.", objective: "Помоги 5 NPC", reward: ["Regeneration", "Heal Burst"], branch: "good" },
  { id: 9, title: "Тёмный путь", dialog: "Ты принял тьму. Сила растёт.", objective: "Уничтожь 10 врагов", reward: ["Double Damage", "Ultimate Shield"], branch: "evil" },
  { id: 10, title: "Битва с самим собой", dialog: "Твоя тень — ты сам. Кто победит?", objective: "Победить зеркального врага", reward: ["Echo Clone", "Quantum Dodge"], branch: null },
  // ... ещё 40 уровней (для примера сократил, добавь сам по шаблону)
  { id: 11, title: "Кристалл времени", dialog: "Время замедляется. Используй это.", objective: "Собери 20 монет за 30 сек", reward: ["Time Slow"] },
  { id: 12, title: "Летающий остров", dialog: "Остров парит. Не упади.", objective: "Долететь до конца", reward: ["Fly Mode"] },
  { id: 13, title: "Легендарный меч", dialog: "Меч зовёт героя.", objective: "Найди меч", reward: ["Laser Eyes"] },
  { id: 14, title: "Босс огня", dialog: "Дракон пробудился.", objective: "Победить дракона", reward: ["Mega Jump"] },
  { id: 15, title: "Тайна подземелья", dialog: "Секреты ждут.", objective: "Найди секретную комнату", reward: ["Wall Walk"] },
  { id: 16, title: "Бесконечная гонка", dialog: "Беги или умри.", objective: "Пробеги 1000 пикселей", reward: ["Dash"] },
  { id: 17, title: "Холодный рассвет", dialog: "Зима пришла.", objective: "Выжить в буре", reward: ["Ice Shield"] },
  { id: 18, title: "Огненный закат", dialog: "Солнце горит.", objective: "Пройди через огонь", reward: ["Fire Aura"] },
  { id: 19, title: "Тень прошлого", dialog: "Прошлое возвращается.", objective: "Уничтожь воспоминания", reward: ["Shadow Teleport"] },
  { id: 20, title: "Финал главы 1", dialog: "Ты нашёл первый осколок. Продолжай.", objective: "Дойди до конца главы", reward: ["Ultimate Power"] },
  // Добавь сам до 50+ — шаблон: { id: N, title: "Название", dialog: "Текст", objective: "Цель", reward: ["Перк1", "Перк2"], branch: null/"good"/"evil" }
];

class StoryManager {
  constructor() {
    this.currentLevel = 0;
    this.shownDialog = false;
    this.dialogTimer = 0;
    this.branch = null;
  }

  getCurrentLevel() {
    return STORY_LEVELS[this.currentLevel];
  }

  advanceLevel(player) {
    if (this.currentLevel < STORY_LEVELS.length - 1) {
      this.currentLevel++;
      this.shownDialog = false;
      this.dialogTimer = 0;
      soundManager.play('level_up', 0.9);
      console.log(`Переход на уровень ${this.currentLevel + 1}: ${this.getCurrentLevel().title}`);
    } else {
      console.log('Сюжет завершён!');
      soundManager.play('win', 1.0);
    }
  }

  showDialog(ctx) {
    const level = this.getCurrentLevel();
    if (!level || this.shownDialog) return;

    ctx.fillStyle = 'rgba(0,0,0,0.7)';
    ctx.fillRect(0, canvas.height - 120, canvas.width, 120);
    ctx.fillStyle = '#ffffaa';
    ctx.font = '20px Arial';
    ctx.fillText(level.title, 50, canvas.height - 90);
    ctx.font = '16px Arial';
    ctx.fillText(level.dialog, 50, canvas.height - 60);
    this.shownDialog = true;
  }

  checkObjective(player, gameState) {
    const level = this.getCurrentLevel();
    if (!level) return;

    // Пример проверки (расширяй под каждую цель)
    if (level.objective.includes('Собери') && player.score >= 5) {
      this.advanceLevel(player);
      level.reward.forEach(perkName => {
        const perk = PERKS_POOL.find(p => p.name === perkName);
        if (perk) perk.apply(player);
        player.perks.push(perkName);
      });
    }
    // Добавь проверки для других целей: убить врагов, время, позиция и т.д.
  }

  update(dt, player, gameState) {
    if (this.dialogTimer < 5) {
      this.dialogTimer += dt;
    } else if (!this.shownDialog) {
      this.showDialog(ctx);
    }
    this.checkObjective(player, gameState);
  }
}

const storyManager = new StoryManager();
